sdk/core/azure-core/src/main/java/com/azure/core/implementation/logging/DefaultLogger.java

```diff
@@ -0,0 +1,376 @@
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void info(String format, Object... args) {
+        logFromFormat(LogLevel.INFORMATIONAL, format, args);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isWarnEnabled() {
+        String logLevelStr = Configuration.getGlobalConfiguration().get(AZURE_LOG_LEVEL);
+        LogLevel currentLogLevel = LogLevel.fromString(logLevelStr);
+        return LogLevel.WARNING.getLogLevel() >= currentLogLevel.getLogLevel();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void warn(String format, Object arg) {
+        logFromFormat(LogLevel.WARNING, format, arg);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param format The formattable message to log.
+     * @param args Arguments for the message. If an exception is being logged, the last argument should be the
+     *     {@link Throwable}.
+     */
+    @Override
+    public void warn(String format, Object... args) {
+        logFromFormat(LogLevel.WARNING, format, args);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isErrorEnabled() {
+        String logLevelStr = Configuration.getGlobalConfiguration().get(AZURE_LOG_LEVEL);
+        LogLevel currentLogLevel = LogLevel.fromString(logLevelStr);
+        return LogLevel.ERROR.getLogLevel() >= currentLogLevel.getLogLevel();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void error(String format, Object arg) {
+        logFromFormat(LogLevel.ERROR, format, arg);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param format The formattable message to log.
+     * @param args Arguments for the message. If an exception is being logged, the last argument should be the
+     *     {@link Throwable}.
+     */
+    @Override
+    public void error(String format, Object... args) {
+        logFromFormat(LogLevel.ERROR, format, args);
+    }
+
+    /**
+     * Format and write the message according to the {@code MESSAGE_TEMPLATE}.
+     *
+     * @param level The level to log.
+     * @param format The log message format.
+     * @param arguments a list of arbitrary arguments taken in by format.
+     */
+    private void logFromFormat(LogLevel level, String format, Object... arguments) {
+        FormattingTuple tp = MessageFormatter.arrayFormat(format, arguments);
+        log(level, tp.getMessage(), tp.getThrowable());
+    }
+
+    /**
+     * Format and write the message according to the {@code MESSAGE_TEMPLATE}.
+     *
+     * @param level log level
+     * @param message The message itself
+     * @param t The exception whose stack trace should be logged
+     */
+    private void log(LogLevel level, String message, Throwable t) {
+        String dateTime = getFormattedDate();
+        String threadName = Thread.currentThread().getName();
+        String levelName = level.name();
+        StringBuilder buf = new StringBuilder(32);
```

Comment:
Any reason to instantiate the `StringBuilder` with an internal capacity of 32? Instead could it be instantiated with a capacity large enough to hold the initial string being written to it?

Reply:
It is inherited from simple logger. 
The message we know definitely will be printed out is:
2020-01-09 12:35 [main] [Informational]

This is 39 character long. I can put 64 instead of 32 characters to minimize capacity increase.